### ISSUE TITLE
Lots of Mineral Exploration mission but rarely any Mining mission

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#628]
**Describe the bug**
- mars-sim  tends to keep starting Mineral Exploration missions

**Expected and unexpected behavior**
- After each or two Mineral Exploration mission(s), the same settlement should consider starting a Mining mission for truly excavating minerals.
 
**Screenshots**
![image](https://user-images.githubusercontent.com/1168584/179076988-6e866223-dd9b-4a2b-84d2-23dde7bb9634.png)

**Specifications**
 - pre 3.4.0 build 6950

**Additional context**
- Mining mission is crucial for the survival of a settlement.
- A mining mission brings in some of those 10 specific types of minerals identified in this [wiki](https://github.com/mars-sim/mars-sim/wiki/Minerals-and-Water-Bound-Ores).
- A lot of these minerals contain a high concentration of iron oxide, which is crucial for manufacturing.

**Comment by mokun** [comment]
@bevans2000 

What's the best way for each settlement to record how many Mineral Exploration have been successfully carried out in the past ?

I know that an explored site is supposed to generate some mining values which will be used by a MiningMeta to calculate probability.

In Mining's, we have getMiningSiteValue method : 
```
	/**
	 * Gets the estimated mineral value of a mining site.
	 * 
	 * @param site       the mining site.
	 * @param settlement the settlement valuing the minerals.
	 * @return estimated value of the minerals at the site (VP).
	 * @throws MissionException if error determining the value.
	 */
	public static double getMiningSiteValue(ExploredLocation site, Settlement settlement) {

		double result = 0D;

		Map<String, Double> concentrations = site.getEstimatedMineralConcentrations();
		Iterator<String> i = concentrations.keySet().iterator();
		while (i.hasNext()) {
			String mineralType = i.next();
			int mineralResource = ResourceUtil.findIDbyAmountResourceName(mineralType);
			double mineralValue = settlement.getGoodsManager().getGoodValuePerItem(mineralResource);
			double concentration = concentrations.get(mineralType);
			double mineralAmount = (concentration / 100D) * MINERAL_BASE_AMOUNT;
			result += mineralValue * mineralAmount;
		}
			
		result = Math.min(100, result);
		return result;
	}
```


But it seems it's always easy to start another Mineral Exploration mission instead of starting a Mining mission.

**Comment by bevans2000** [comment]
We should change the landmarks.xml so it is possible to pre-load fully explored sites at the start. This will increase the change of a Mining mission starting

**Comment by Urwumpe** [comment]
Can we include this as a CLI option? That could be useful for testing purposes. Or a cheat command in the command window to make all sites explored.

**Comment by bevans2000** [comment]
Console command, good idea. Probably only triggered in expert mode

**Comment by mokun** [comment]
> We should change the landmarks.xml so it is possible to pre-load fully explored sites at the start. This will increase the change of a Mining mission starting

So the question is, how many unexplored sites would be necessary in triggering a mining mission successfully ? 

We don't need to have 10 exploration missions done before we start a mining mission, right ?

But right now the ratio is about 10:1.

I would think, immediately right after an exploration mission (with a couple of waypoints explored) is done, one can start two or three mining missions already. 

So if it makes sense this way, then we'll need to figure out a way to change how MiningMeta calculate probability.

**Comment by bevans2000** [comment]
I'm going to add the capability to create new ```ExploredLocation``` via the Console. 
An explored site needs an improvement score of 10 before it can be mined. That is limiting the Mining probability.
We should also change the ownership of explored location from Settlement to ReportingAuthority as it will allow them to be shared between collaborating Settlements. In addition, the locations could be unclaimed without a Reporting Authority; these would be claimed as part of the reservation in the Mining activity.

**Comment by mokun** [comment]
> In addition, the locations could be unclaimed without a Reporting Authority; these would be claimed as part of the reservation in the Mining activity.

Ah... I can see that staking one's claim would be an touchy subject and an interesting one on Mars. 

At some points, a central authority will need to be formed in processing and arbitrating those prospecting/mining claims.

**Comment by mokun** [comment]
> I'm going to add the capability to create new ExploredLocation via the Console.

We already have the `/ep' command to look up the explored locations under the settlement.

See below: 

```
00-Adir-01:140 MarsNet >  ​/?

Control Keys :
 * tab - Auto completes the input; for commands (long & short) and arguments
 * up - Goes back in the history
 * down - Goes forward in the history

Console Commands (* <short command> <long command>)
Group : Common
 * /xp expert              * /-  intro
 * /?  help                * /rt repeat

Group : Simulation
 * /c  connect             * /ms mission summary
 * /d  date                * /p  pause
 * /di distance            * /mp planning
 * /el elevation           * /sv save
 * /ev events              * /se settlements
 * /ep explored            * /sp speed
 * /lo log                 * /st stop

More detailed help by using > help <long command>
00-Adir-01:635 MarsNet >  ​/ep
          Location |           Settlement |   Status | Reviews | Highest
------------------------------------------------------------------------

00-Adir-01:649 MarsNet >  ​
```

But if we are going to create a *centralized* database that is not tied to specifically one settlement, then we'll need to change that.

**Comment by bevans2000** [comment]
Yes I know. I'm going to add the capability to the existing command.

**Comment by mokun** [comment]
> these would be claimed as part of the reservation in the Mining activity.

So at this point, let's say we can make the claiming process to be as objective as possible by including 3 things : 
 
1. Scheduling the computing center to run some computer simulations as proof that the intention is to look at that location (before and/or after being explored) 
2. Being there physically at a certain time
3. Taking back some rock samples at that location.

**Comment by bevans2000** [comment]
Last night I added the ability to create new ```ExploredLocations``` via the Console. To do this you just run ```explored add``` from the Console. This will prompt you for the coordinates of the new location. The Console will check that there are actually minerals present at the location as a pre-req.  If minerals are present then the location is created with a improvement score of 10 and the status of 'explored'. This allows it to be a candidate for Mining missions.

**Comment by mokun** [comment]
> will check that there are actually minerals present at the location as a pre-req. 

ic.

> If minerals are present then the location is created with a improvement score of 10 and the status of 'explored'. This allows it to be a candidate for Mining missions.

I know we have 3 png image files that are being used for discovering minerals. 

But we certain want to go beyond just using those old data sets. 

Plus, we have said we want to use instruments for discovering minerals.

Btw, in this recent [commit](https://github.com/mars-sim/mars-sim/commit/d4cee9228029872ee3abc90c65e9b50d93437202) I changed the probability calculation for finding various rock types and refining mineral concentration. Just so you know.

In `ExploreSite` task as follows : 
```
	// Data members
	private double totalCollected = 0;
	private double numSamplesCollected = AVERAGE_ROCK_SAMPLES_COLLECTED_SITE / AVERAGE_ROCK_SAMPLE_MASS;
	
	// Future: should keep track of the actual total exploring site time and use it below.
	// The longer it stays, the more samples are collected and better the mining estimation
	private double chance = numSamplesCollected * Exploration.EXPLORING_SITE_TIME / 8000.0;
```

Question : Can we record the *actual* site time at each waypoint ? 

Currently, we set EXPLORING_SITE_TIME as to 500 millisols by default. It should be a variable and it should affect how well you find the rock types.

**Comment by bevans2000** [comment]
Yes with the new design each Eva phase  an have a different time. There is an abstract method overriden that provides the length of the EVA. At the moment each subclass does you a fixed value but it could be based on the current site

**Comment by mokun** [comment]
After some changes, things seem to get better but not perfect yet, in this simulation on sol 5, the StarCity have done 4 Mineral Exploration, 1 regolith prospecting and 2 Mining missions. 

![image](https://user-images.githubusercontent.com/1168584/180660127-1daea757-14eb-480b-bfc6-9f88c7039729.png)

For Mining #14, however, the amount of minerals uncovered is too little. Out of the 10 minerals available, it makes sense that certain types are found. But together, they are less than 20 kg. Does it worth the mission ?

![image](https://user-images.githubusercontent.com/1168584/180659958-918b4a8b-5aca-4466-8288-950556543742.png)

Q : why the schedule did not record the exploring sites and excavating minerals in Mining #14, even though it does show it picks up those 20 kg minerals ?

![image](https://user-images.githubusercontent.com/1168584/180660093-0d43ce89-1b9c-4e8e-b958-227141948e4f.png)

So far, this mining mission shows no sites visited in the Navigation tab.

//////////////////////////////////////

On the other hand, the on-going Mining #33 is different. 

It shows in the Navigation tab it has visited 3 sites. 

![image](https://user-images.githubusercontent.com/1168584/180660910-39f49991-9bec-4f38-83db-92e73be583ce.png)

On the schedule, however, it shows NO site are being visited nor excavating. It also has `EVA - Not enough sunlight` in the Phase.

![image](https://user-images.githubusercontent.com/1168584/180661084-1d3f37d7-411b-4192-b19e-b725e95e37fc.png)

**Comment by bevans2000** [comment]
The last commit now allow you to configure the minimum sunlight for EVA. So if you set it to zero; then EVAs will never be blocked due to poor light hence allowing us to discount the sunlight in the above cases.

**Comment by mokun** [comment]
> configure the minimum sunlight for EVA. So if you set it to zero; then EVAs will never be blocked due to poor light hence allowing us to discount the sunlight in the above cases.

So the min-eva-light value is 7. How often would it stay between 0 and 7 ? 

I thought it would be a very quick transition. May be in a few millisols it would go from 7 to 0. If that's the case, would it help ?



 

I suppose this is a stop gap measure for now

**Comment by bevans2000** [comment]
Not exactly because there will be a large period when it is zero because the sun is below the horizon. 
There is an odd behaviour where the hijacked simulation XML gets replaced automatically. I think I've found why its doing it. It seems we clear the directory of the version changes but its not needed because the MD5 check catches any changed files.
I don't see it as a stop gap because we may want to change the value anyway.

**Comment by mokun** [comment]
> there will be a large period when it is zero because the sun is below the horizon.

oic.

> There is an odd behaviour where the hijacked simulation XML gets replaced automatically.

hijacked ?

> I don't see it as a stop gap because we may want to change the value anyway.

But we want to be able to set up a schedule for all normal settlement EVAs and mission EVAs in future, right ?

May be there's a better way of putting it. 

A bigger picture is that we want a master schedule, something @Urwumpe has suggested in another thread.

Say if NASA were to do it their usual way, the mission control would impose their version of the daily schedule of their astronaut, such as when to exercise, rest and go out for an EVA, etc. 

But of course, just like in the real world, things aren't perfect and emergency arises. What settlers will end up doing will differ.

So we could let player set up a daily schedule programmatically by dragging and dropping tasks into a blank schedule, even using Swing. 

But may be you can create a command in text console for players to create this master schedule, do you think so ?

**Comment by bevans2000** [comment]
Yes it's possible. The Shift logic should be considered as well.

**Comment by mokun** [comment]
We'll delay the refinement on this issue until later.

**Comment by mokun** [comment]
It dawns on me the other day that we enable settlers to ***FIRST*** and foremost explore the area right where the settlement is located at. 

Normally one of the criteria in selecting a landing site in close proximity to various Resource point of Interest and Science point of Interest (POI), right ? 

Can we add one or two of areological features for each settlement ? 

And we could rework ExploreSite to have a ExploreSiteMeta so that settlers can come out and even practice their techniques first right outside their base. 

It's not logical to make them go on far-away missions at the start of the sim. 

We could have them mine minerals at the base and do some outside field study such as checking out those areological interest.
